### DISCUSSION TITLE
How to best use the parent-child ref in mongoose schema

### DISCUSSION THREAD

**Discussion by bwgjoseph** [feathersjs/feathers#2348] [Q&A]
Hey, wanna check with you guys on how do you guys do parent-child ref for mongoose r/s. I know that using ref in the schema, you can do it but all the example shown are quite straight-forward like such https://mongoosejs.com/docs/populate.html

However, it isn't always the case where the parent and child are created together, and for those that do not, there would have an additional query to update the parent whenever there is a new child. So for example, `post and comments`, the user creating a post and other users creating a comment on the post itself is a separate action altogether.

```js
// userA create post, in the schema, a comment ref is defined
service('post').create({ name: 'new post' }) // assuming return id 12345

// userB commented on the post
service('comment').create({ description: 'hello from a', postId: 12345 }) // assuming return id 11111
// a patch to post 12345 is required
service('post).patch('12345', { $push: { commentIds: '11111' } });

// userC commented on the post
service('comment').create({ description: 'hello from b', postId: 12345 }) // assuming return id 22222
// a patch to post 12345 is required
service('post).patch('12345', { $push: { commentIds: '22222' } });
```

To further illustrate the possible scenario is where both `post` and `comment` will be published via websocket to users, so in the scenario above, we would have...

```js
// post 12345 created event
// comment 11111 created event
// post 12345 patched event
// comment 22222 created event
// post 12345 patched event
```

Is this how usually you guys would have done it? Because each new `comment` would have been emitted to client but with that, also comes along a `patched post` event to client which might be quite redundant. 

What is the best course of action for this kind of scenario? Could it be for this scenario, using `_patch` is better since it does not publish events?

Thanks!

**Comment by tvld** [comment]
I am a big fan of using [fastJoin](https://hooks-common.feathersjs.com/hooks.html#fastjoin) to populate a document with it's children. It's fast, steady and although it took me a moment to figure it out, now I do not even know why I thought it was complex once)))

As for the channel publishing, I rewrote the publish in ``channels.js`` in such a way that it only pusblishes to other users:
```js
  // find what services (by name) we will publish as in our case we will not publish all 
  const chnAllMembers = app.get('publishChannels').allMembers // services to publish for members of account
  const chnOthers = app.get('publishChannels').othersOnly // services for non-members
  const chnAll =[...chnAllMembers, ...chnOthers] // all services

  // Share all events from services that have an *accountId* on the account channels
  // eslint-disable-next-line no-unused-vars
  app.publish((data, context) => {
    const path = context.path
    // check if we have an account channel to publish to
    if (chnAll.includes(path) && data._id && data.accountId) {
      const newChn = Object.assign({}, app.channel(accountChn(data.accountId)))

      if (chnOthers.includes(path) && context.params.user) {
        const userId = context.params.user._id
        // console.log(`>>> ${accountChn(data.accountId)} ${connections.length}`)
        newChn.connections = newChn.connections.filter(connection => {
          const connectionUserId = connection.user._id
          return connectionUserId !== userId
        })
      }
      if (newChn.connections.length  > 0) {
        console.log(`>> Publishing "${path}" events on channel "${accountChn(data.accountId)}" with ${newChn.connections.length} connections`)
      }
      return newChn
    }
  })
```

**Reply by bwgjoseph** [reply]
Hi @tvld,

Thanks for your thoughts. I have used `fastJoin` previously but removed it subsequently as it is quite difficult for other teams using other languages like `Java` to dynamically determine if a response has `joined` data or not since Java needs the POJO to be explicitly declared upfront where javascript is more flexible in the data structure. It causes some confusion but it has its uses, may be able to re-visit this in the future. But this is not the main point.

I wanted to understand a little more about how people deal with such a scenario, and how they deal with it. There could be ways that I haven't thought so before but with some more research, this [post](https://stackoverflow.com/questions/32126367/is-it-necessary-hold-references-to-children-in-parent-document-mongooose) in StackOverflow seem to cover it (mostly), and it basically boils down to your design choice and what you choose to use since each has its pros and cons.

In this case, most of my design is based on `backref to parent on child` which mostly works out ok at the moment
